src/compute-client/src/controller/instance.rs

```diff
@@ -1178,6 +1180,11 @@ where
         let mut compute_dependencies = BTreeMap::new();
 
         for &id in dataflow.source_imports.keys() {
+            // TODO(ct): Hack because we can use our output as an input.
+            if dataflow.sink_exports.contains_key(&id) {
+                continue;
+            }
```

Comment:
Great questions, this is something I haven't hashed out yet, and for sure it's probably slightly wrong at the moment (certainly wrong for the restart case). Gonna split this out into a slack thread.

Reply:
(Didn't get a chance to make this thread today, will do tomorrow)

Reply:
It ended up not as complicated as I thought.

- The CT starts at some start_ts/restart_ts, call it `T`
- It renders a persist_source with snapshot=exclude for each input with `as_of == T`
  - This guarantees that we see any diffs at times after `T`: i.e. the first thing coming out of here is `T+1`
  - To be allowed to do this, we need the shard since to be at most `T`
- It renders a persist_source with snapshot=include for each reference with `as_of == T`
  - The first thing coming out of here is `T`
  - As usual, we need the shard since here to be at most `T`
- This ends up writing at `T+1`
  - Notably, this is different than an MV, which holds back the input sinces to at most `T` and writes at `T`

> I assume it must be the latter, as the former wouldn't work when we only have read holds at as_of. If this is the case, the fix for the scheduling issue would be to create the storage collection for the output with frontiers (as_of, as_of + 1), with the as_of time immediately readable as an empty collection.

Oh that's clever! Does anything else work like that?

> I imagine we'd have to do this anyway, to ensure the CT dataflow can even start up. Though it seems like we're not doing it right now, so I'm clearly missing something...

This lives in rendering, where you mentioned you haven't gotten to yet. We carefully defined the CT's behavior to only write at times that an input changed. So as soon as it finds out that there is nothing in `ct_times` before `T+1`, it does a big empty compare and append of `[0, T+1)`

Reply:
> Oh that's clever! Does anything else work like that?

Not that I'm aware of. In fact we might need to add a storage controller API that allows you to create a collection that immediately has readable times. I think `create_collections` always initializes the read frontier to the given as-of but leaves the write frontier at zero.

> We carefully defined the CT's behavior to only write at times that an input changed. So as soon as it finds out that there is nothing in `ct_times` before `T+1`, it does a big empty compare and append of `[0, T+1)`

Okay, that checks out then. We'd basically make the append in the controller by creating the output shard with a write frontier of `T + 1` instead.

Reply:
> > Oh that's clever! Does anything else work like that?
> 
> Not that I'm aware of. In fact we might need to add a storage controller API that allows you to create a collection that immediately has readable times. I think `create_collections` always initializes the read frontier to the given as-of but leaves the write frontier at zero.
> 
> > We carefully defined the CT's behavior to only write at times that an input changed. So as soon as it finds out that there is nothing in `ct_times` before `T+1`, it does a big empty compare and append of `[0, T+1)`
> 
> Okay, that checks out then. We'd basically make the append in the controller by creating the output shard with a write frontier of `T + 1` instead.

Oof, I haven't tried it yet but the storage controller API change sounds messy. I don't have any other ideas, though...

Reply:
If we can't create the shard with a write frontier of our choosing directly, we can maybe convince the storage controller to do an initial append after its creation? Though I assume that gets messy too, especially when envd crashes before managing to perform the append.

Alternatively, we could do the initial append in the CT sink too. We'd just have to do it before waiting on the `start_signal`, to ensure it happens even when the dataflow hasn't been scheduled yet.

Reply:
That worked beautifully :100:

Reply:
Neat :)